src/_pytest/faulthandler.py

```diff
@@ -25,92 +25,72 @@ def pytest_addoption(parser: Parser) -> None:
 def pytest_configure(config: Config) -> None:
     import faulthandler
 
-    if not faulthandler.is_enabled():
-        # faulthhandler is not enabled, so install plugin that does the actual work
-        # of enabling faulthandler before each test executes.
-        config.pluginmanager.register(FaultHandlerHooks(), "faulthandler-hooks")
-    else:
-        # Do not handle dumping to stderr if faulthandler is already enabled, so warn
-        # users that the option is being ignored.
-        timeout = FaultHandlerHooks.get_timeout_config_value(config)
-        if timeout > 0:
-            config.issue_config_time_warning(
-                pytest.PytestConfigWarning(
-                    "faulthandler module enabled before pytest configuration step, "
-                    "'faulthandler_timeout' option ignored"
-                ),
-                stacklevel=2,
-            )
-
-
-class FaultHandlerHooks:
-    """Implements hooks that will actually install fault handler before tests execute,
-    as well as correctly handle pdb and internal errors."""
-
-    def pytest_configure(self, config: Config) -> None:
-        import faulthandler
+    stderr_fd_copy = os.dup(get_stderr_fileno())
+    config._store[fault_handler_stderr_key] = open(stderr_fd_copy, "w")
+    faulthandler.enable(file=config._store[fault_handler_stderr_key])
 
-        stderr_fd_copy = os.dup(self._get_stderr_fileno())
-        config._store[fault_handler_stderr_key] = open(stderr_fd_copy, "w")
-        faulthandler.enable(file=config._store[fault_handler_stderr_key])
 
-    def pytest_unconfigure(self, config: Config) -> None:
-        import faulthandler
+def pytest_unconfigure(config: Config) -> None:
+    import faulthandler
 
-        faulthandler.disable()
-        # close our dup file installed during pytest_configure
-        # re-enable the faulthandler, attaching it to the default sys.stderr
-        # so we can see crashes after pytest has finished, usually during
-        # garbage collection during interpreter shutdown
+    faulthandler.disable()
+    # Close the dup file installed during pytest_configure.
+    if fault_handler_stderr_key in config._store:
         config._store[fault_handler_stderr_key].close()
         del config._store[fault_handler_stderr_key]
-        faulthandler.enable(file=self._get_stderr_fileno())
+    # Re-enable the faulthandler, attaching it to the default sys.stderr
```

Comment:
So IIUC this will enable faulthandler for the host process after pytest is done even if it was not enabled before, right? I think maybe we only want to do this part if faulthandler was enabled before `pytest_configure`, just so we leave things as we got them?

Reply:
Definitely, good catch. 👍

Reply:
Hmmm on second thought not sure. 

The comment says: "so we can see crashes after pytest has finished", so this will not happen when executing `pytest` normally, where `faulthandler` was not enabled before.

I suggest we leave it like this, seems the more useful setting. What do you think?

Reply:
I agree it's useful, but IMO "after pytest has finished" is out of our domain. We should perhaps encourage users to run pytest with `-X dev`, maybe even do it automatically when the `pytest` entry point is used (I might open an issue to discuss this :thinking: ). But if someone is using `pytest.main()` I think we shouldn't touch it (that is, restore to previous value).

Reply:
> I agree it's useful, but IMO "after pytest has finished" is out of our domain.

I personally think practicability wins here, but no worries, I will apply the requested change.

Reply:
Hmm I realized `test_crash_near_exit` is meant to test the behavior of catching crashes during interpreter shutdown:

```python
    def test_crash_near_exit(pytester: Pytester) -> None:
        """Test that fault handler displays crashes that happen even after
        pytest is exiting (for example, when the interpreter is shutting down)."""
        pytester.makepyfile(
            """
        import faulthandler
        import atexit
        def test_ok():
            atexit.register(faulthandler._sigabrt)
        """
        )
        result = pytester.runpytest_subprocess()
        result.stderr.fnmatch_lines(["*Fatal Python error*"])
```

Applying the requested change makes the test fail (as expected). Really think this is the way to go @bluetech?

Reply:
Yes -- IMO like campers say, "leave it as you found it". Given that `pytest.main` is a public API, we shouldn't assume that we are the complete owners of the process after we finish. Ideally we would not even assume it *while* we are running but that is not really possible unfortunately (due to things like replacing `sys.stdout` and other global resources).

On the other hand when we are run from the cli then we *can* assume that so we can enable whatever we want then, but needs some discussion.

Reply:
Fair enough, done. 👍